atc/component/scheduler_test.go

```diff
@@ -0,0 +1,158 @@
+package component_test
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	"code.cloudfoundry.org/clock"
+	"code.cloudfoundry.org/clock/fakeclock"
+	"code.cloudfoundry.org/lager/lagertest"
+	"github.com/concourse/concourse/atc/component"
+	"github.com/concourse/concourse/atc/component/cmocks"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"github.com/tedsuo/ifrit"
+)
+
+func TestScheduler(t *testing.T) {
+	suite.Run(t, &SchedulerSuite{
+		Assertions: require.New(t),
+	})
+}
+
+type SchedulerSuite struct {
+	suite.Suite
+	*require.Assertions
+
+	clock *fakeclock.FakeClock
+}
+
+func (s *SchedulerSuite) SetupTest() {
+	s.clock = fakeclock.NewFakeClock(time.Now())
+	component.Clock = s.clock
+}
+
+func (s *SchedulerSuite) TearDownTest() {
+	component.Clock = clock.NewClock()
+}
+
+func (s *SchedulerSuite) TestEndToEnd() {
+	interval := 30 * time.Second
+	componentName := "some-component"
+
+	mockComponent := new(cmocks.Component)
+	mockComponent.On("Name").Return(componentName)
+	mockComponent.On("Interval").Return(interval)
+
+	mockBus := new(cmocks.NotificationsBus)
+
+	ranPeriodically := make(chan context.Context)
+	ranImmediately := make(chan context.Context)
+
+	mockSchedulable := schedulable{
+		runPeriodically: func(ctx context.Context) {
+			ranPeriodically <- ctx
+		},
+		runImmediately: func(ctx context.Context) {
+			ranImmediately <- ctx
+		},
+	}
+
+	scheduler := &component.Scheduler{
+		Logger:      lagertest.NewTestLogger("test"),
+		Interval:    interval,
+		Component:   mockComponent,
+		Bus:         mockBus,
+		Schedulable: mockSchedulable,
+	}
+
+	notifications := make(chan bool, 1)
+
+	var process ifrit.Process
+	s.Run("listens for component notifications on start", func() {
+		mockBus.On("Listen", componentName).Return(notifications, nil)
+
+		process = ifrit.Background(scheduler)
+		select {
+		case <-process.Ready():
+		case err := <-process.Wait():
+			s.Failf("process exited early", "error: %s", err)
+		}
+
+		mockBus.AssertCalled(s.T(), "Listen", componentName)
+	})
+
+	defer func() {
+		process.Signal(os.Interrupt)
+		<-process.Wait()
+	}()
+
+	s.Run("runs periodically on component interval", func() {
+		s.clock.WaitForWatcherAndIncrement(interval)
+		<-ranPeriodically
```

Comment:
Yeah, these test situations are kind of annoying since both approaches have downsides.

I stopped using `Eventually` even in Ginkgo/Gomega because it ended up causing way too many CI flakes that were just caused by the worker being under more load than usual.

Nowadays I prefer to accept the risk of the test hanging indefinitely - usually those failures don't escape developer machines, and if they make it to CI they tend to motivate people to fix the tests. 😄 I think that's the better tradeoff since it prevents us from becoming numb to flakiness.